pkg/listwatch/listwatch.go

```diff
@@ -0,0 +1,179 @@
+		}
+		items, err := meta.ExtractList(list)
+		if err != nil {
+			return nil, err
+		}
+		metaObj, err := meta.ListAccessor(list)
+		if err != nil {
+			return nil, err
+		}
+		for _, item := range items {
+			l.Items = append(l.Items, runtime.RawExtension{Object: item.DeepCopyObject()})
+		}
+		resourceVersions = append(resourceVersions, metaObj.GetResourceVersion())
+	}
+	// Combine the resource versions so that the composite Watch method can
+	// distribute appropriate versions to each underlying Watch func.
+	l.ListMeta.ResourceVersion = strings.Join(resourceVersions, "/")
+	return &l, nil
+}
+
+// Watch implements the ListerWatcher interface.
+// It returns a watch.Interface that combines the output from the
+// watch.Interface of every cache.ListerWatcher into a single result chan.
+func (mlw multiListerWatcher) Watch(options metav1.ListOptions) (watch.Interface, error) {
+	resourceVersions := make([]string, len(mlw))
+	// Allow resource versions to be "".
+	if options.ResourceVersion != "" {
+		rvs := strings.Split(options.ResourceVersion, "/")
+		if len(rvs) != len(mlw) {
+			return nil, fmt.Errorf("expected resource version to have %d parts to match the number of ListerWatchers", len(mlw))
+		}
+		resourceVersions = rvs
+	}
+	return newMultiWatch(mlw, resourceVersions, options)
+}
+
+// multiWatch abstracts multiple watch.Interface's, allowing them
+// to be treated as a single watch.Interface.
+type multiWatch struct {
+	result   chan watch.Event
+	stopped  chan struct{}
+	stoppers []func()
+}
+
+// newMultiWatch returns a new multiWatch or an error if one of the underlying
+// Watch funcs errored. The length of []cache.ListerWatcher and []string must
+// match.
+func newMultiWatch(lws []cache.ListerWatcher, resourceVersions []string, options metav1.ListOptions) (*multiWatch, error) {
+	var (
+		result   = make(chan watch.Event)
+		stopped  = make(chan struct{})
+		stoppers []func()
+		wg       sync.WaitGroup
+	)
+
+	wg.Add(len(lws))
+
+	for i, lw := range lws {
+		o := options.DeepCopy()
+		o.ResourceVersion = resourceVersions[i]
+		w, err := lw.Watch(*o)
+		if err != nil {
+			return nil, err
+		}
+
+		go func() {
+			defer wg.Done()
+
+			for {
+				event, ok := <-w.ResultChan()
+				if !ok {
+					return
+				}
+
+				select {
+				case result <- event:
+				case <-stopped:
+					return
+				}
+			}
+		}()
+
+		stoppers = append(stoppers, w.Stop)
+	}
+
+	// result chan must be closed,
+	// once all event sender goroutines exited.
+	go func() {
+		wg.Wait()
+		close(result)
+	}()
+
+	return &multiWatch{
+		result:   result,
+		stoppers: stoppers,
+		stopped:  stopped,
+	}, nil
+}
+
+// ResultChan implements the watch.Interface interface.
+func (mw *multiWatch) ResultChan() <-chan watch.Event {
+	return mw.result
+}
+
+// Stop implements the watch.Interface interface.
+// It stops all of the underlying watch.Interfaces and closes the backing chan.
+// Can safely be called more than once.
+func (mw *multiWatch) Stop() {
+	select {
```

Comment:
There is a race condition here, which can happen when Stop gets called reentrantly. stop methods can be called multiple times. Not sure if this is a problem with the sub-watchers.